flask-server/main.py

```diff
@@ -421,11 +406,73 @@ def fetch_bookmarks(clink_id):
 
 @app.route('/apis/fetch-username/<string:user_id>', methods=['GET'])
 def fetch_username(user_id):
-    query = datastore_client.query(kind='user')
     key = datastore_client.key('user', int(user_id))
-    user = list(query.add_filter('__key__', '=', key).fetch(limit=1))[0]
+    user = list(datastore_client.query(kind='user').add_filter('__key__', '=', key).fetch(limit=1))[0]
     return user['username'], 200
 
+# edit bookmark api
+@app.route('/apis/edit-bookmark', methods=['POST'])
+def edit_bookmark():
+    resp_token = decode_auth_token(request.json['Authorization'])
+    clink_id = request.json['clinkId']
+
+    if is_valid_instance(resp_token) and has_write_access(clink_id, resp_token):
+        link = request.json['link']
+        title = request.json['title']
+        description = request.json['description']
+
+        key = datastore_client.key('bookmark', int(request.json['bookmarkId']))
+        bookmark = datastore_client.get(key)
+        
+        if link:
+            bookmark['link'] = link
+        if title:
+            bookmark['title'] = title
+        if description: 
+            bookmark['description'] = description
+
+        datastore_client.put(bookmark)
+
+        return bookmark_entity_to_return(bookmark), 200
+    else: 
+        response_object = {
+            'status': 'fail',
+            'message': 'Invalid JWT. Failed to edit bookmark.'
+        }
+        return response_object, 401
+
+# edit clink api
+@app.route('/apis/edit-clink', methods=['POST'])
+def edit_clink():
+    resp_token = decode_auth_token(request.json['Authorization'])
+    clink_id = request.json['clinkId']
+
+    if is_valid_instance(resp_token) and has_write_access(clink_id, resp_token):
+        title = request.json['title']
+
+        key = datastore_client.key('clink', int(clink_id))
+        clink = datastore_client.get(key)
+        
+        if title:
+            clink['title'] = title
+
+            datastore_client.put(clink)
+
+        return clink_entity_to_return(clink), 200
+    else: 
+        response_object = {
+            'status': 'fail',
+            'message': 'Invalid JWT. Failed to edit clink.'
+        }
+        return response_object, 401
+
+# write access api
+def has_write_access(clink_id, user_id):
+    access = True if clink_id == "All" else list(datastore_client.query(kind='user_write_map').add_filter('clink_id', '=', int(clink_id)).add_filter('user_id', '=', int(user_id)).fetch(limit=1))[0]
+    if access:
```

Comment:
How should check write access for the 'All' clink?

Reply:
I'd probably just return true if the id it 'All' since you should never be able to access someone else's 'All'

but you should only be checking write access for editable clinks, and 'All' isn't an editable clink. Basically this should never be called with 'All'